src/actions/requests.rs

```diff
@@ -852,27 +854,78 @@ fn location_from_racer_match(a_match: racer::Match) -> Option<Location> {
 }
 
 lazy_static! {
+    static ref NUM_THREADS: usize = ::num_cpus::get();
+
+    /// Duration of work after which we should warn something is taking a long time
+    static ref WARN_TASK_DURATION: Duration = *DEFAULT_REQUEST_TIMEOUT * 5;
+
+    /// Current work descriptions active on the work pool
+    static ref WORK: Mutex<Vec<&'static str>> = Mutex::new(vec![]);
+
     /// Thread pool for request execution allowing concurrent request processing.
     static ref WORK_POOL: rayon::ThreadPool = rayon::ThreadPool::new(
         rayon::Configuration::default()
             .thread_name(|num| format!("request-worker-{}", num))
-            // panic details will be on stderr, otherwise ignore the work panic as it
-            // will already cause a mpsc disconnect-error & there isn't anything else to log
-            .panic_handler(|_| {})
+            .num_threads(*NUM_THREADS)
     ).unwrap();
 }
 
 /// Runs work in a new thread on the `WORK_POOL` returning a result `Receiver`
 /// Panicking work will receive `Err(RecvError)` / `Err(RecvTimeoutError::Disconnected)`
-pub fn receive_from_thread<T, F>(work_fn: F) -> mpsc::Receiver<T>
+pub fn receive_from_thread<T, F>(work_fn: F, description: &'static str) -> mpsc::Receiver<T>
 where
     T: Send + 'static,
-    F: FnOnce() -> T + Send + 'static,
+    F: FnOnce() -> T + Send + panic::UnwindSafe + 'static,
 {
     let (sender, receiver) = mpsc::channel();
+
+    {
+        let mut work = WORK.lock().unwrap();
+        if work.len() >= *NUM_THREADS {
+            // there are already N ongoing tasks, that may or may not have timed out
+            // don't add yet more to the queue fail fast to allow the work pool to recover
+            warn!(
+                "Could not start `{}` as at work capacity, {:?} in progress",
+                description,
+                *work,
+            );
+            return receiver;
+        }
+        if work.iter().filter(|desc| *desc == &description).count() >= (*NUM_THREADS / 2).max(1) {
```

Comment:
Could you factor out `(*NUM_THREADS / 2).max(1)` into another static?

Reply:
Sure